spec/dynamoid/transaction_write/upsert_spec.rb

```diff
@@ -0,0 +1,250 @@
+            txn.upsert! obj5
+          end
+          obj5_found = klass_with_composite_key.find(obj5.id, range_key: 5)
+          expect(obj5_found).to eql(obj5)
+          expect(obj5_found.name).to eql("five")
+        end
+        it 'requires hash key in class' do
+          expect {
+            Dynamoid::TransactionWrite.execute do |txn|
+              txn.upsert! klass_with_composite_key, {name: "threethree"}
+            end
+          }.to raise_exception(Dynamoid::Errors::MissingHashKey)
+        end
+        it 'requires hash key in instance' do
+          expect {
+            Dynamoid::TransactionWrite.execute do |txn|
+              txn.upsert! klass_with_composite_key.new(name: "threethree")
+            end
+          }.to raise_exception(Dynamoid::Errors::MissingHashKey)
+        end
+        it 'requires range key in class' do
+          expect {
+            Dynamoid::TransactionWrite.execute do |txn|
+              txn.upsert! klass_with_composite_key, {id: "bananas", name: "threethree"}
+            end
+          }.to raise_exception(Dynamoid::Errors::MissingRangeKey)
+        end
+        it 'requires range key in instance' do
+          expect {
+            Dynamoid::TransactionWrite.execute do |txn|
+              txn.upsert! klass_with_composite_key.new(id: "bananas", name: "threethree")
+            end
+          }.to raise_exception(Dynamoid::Errors::MissingRangeKey)
+        end
+      end
+
+      it 'updates timestamps of instance' do
+        klass.create_table
+        obj1 = klass.create!(name: "one", created_at: Time.now - 48.hours, updated_at: Time.now - 24.hours)
+        obj1.name = "oneone"
+        Dynamoid::TransactionWrite.execute do |txn|
+          txn.upsert! obj1
+        end
+        obj1_found = klass.find(obj1.id)
+        expect(obj1_found.created_at.to_f).to be < (Time.now - 47.hours).to_f
+        expect(obj1_found.updated_at.to_f).to be_within(1.seconds).of Time.now.to_f
+      end
+
+      it 'updates timestamps by class when existing' do
+        klass.create_table
+        obj3 = klass.create!(name: "three", created_at: Time.now - 48.hours, updated_at: Time.now - 24.hours)
+        Dynamoid::TransactionWrite.execute do |txn|
+          txn.upsert! klass, {id: obj3.id, name: "threethree"}
+        end
+        obj3_found = klass.find(obj3.id)
+        expect(obj3_found.created_at.to_f).to be < (Time.now - 47.hours).to_f
+        expect(obj3_found.updated_at.to_f).to be_within(1.seconds).of Time.now.to_f
+      end
+
+      it 'updates timestamps by class when not existing' do
+        klass.create_table
+        obj3_id = SecureRandom.uuid
+        Dynamoid::TransactionWrite.execute do |txn|
+          txn.upsert! klass, {id: obj3_id, name: "threethree"}
+        end
+        obj3_found = klass.find(obj3_id)
+        expect(obj3_found.created_at).to be_nil
+        expect(obj3_found.updated_at.to_f).to be_within(1.seconds).of Time.now.to_f
+      end
+
+      context 'validation' do
```

Comment:
The `.upsert` method `doesn't run validations and callbacks`. So at least this behaviour shouldn't be tested.

Reply:
So there is no `upsert!` method (with `!`), only `upsert`.